module-apps/application-call/windows/CallWindow.cpp

```diff
@@ -299,34 +299,61 @@ namespace gui
 
     void CallWindow::handleAudioEvent(const audio::Event &event)
     {
-
-        switch (event.getType()) {
-        case audio::EventType::BlutoothHFPDeviceState:
-        case audio::EventType::BlutoothHSPDeviceState:
+        const auto eventType = event.getType();
+        switch (eventType) {
+        case audio::EventType::BluetoothHFPDeviceState:
+        case audio::EventType::BluetoothHSPDeviceState:
         case audio::EventType::JackState: {
             if (event.getDeviceState() == audio::Event::DeviceState::Connected) {
-                devices_connected.insert(event.getType());
-                changeSpeakerIconIfNeeded();
+                handleConnectAudioEvent(eventType);
             }
             else {
-                devices_connected.erase(event.getType());
-                if (not devices_connected.empty()) {
-                    changeSpeakerIconIfNeeded();
-                }
+                handleDisconnectAudioEvent(eventType);
             }
             break;
         }
+        case audio::EventType::CallLoudspeakerOn: {
+            isLouderSpeakerON = true;
+            break;
+        }
+        case audio::EventType::CallLoudspeakerOff: {
+            isLouderSpeakerON = false;
+            break;
```

Comment:
These cases should be passed to the presenter and handled there too:) Try to avoid adding boolean variables holding different kinds of state, and instead implement logic using proper state machines or use already existing mechanisms to do the job:)